Model/Api.php

```diff
@@ -1833,11 +1835,24 @@ public function kickoffParallelCampaigns(Contact $contact, &$campaignIds = [])
             return $this;
         }
 
+        if (!$allowFork) {
+            return $this;
+        }
+
         // Commit any MySQL changes and close the connection/s to prepare for a process fork.
+        /** @var Connection $connection */
         $connection = $this->em->getConnection();
-        if ($connection->isConnected()) {
-            if ($connection->isTransactionActive()) {
-                $connection->commit();
+        if ($connection) {
+            while (0 !== $connection->getTransactionNestingLevel()) {
+                // Check for RollBackOnly to avoid exceptions.
+                if (!$connection->isRollbackOnly()) {
+                    // Follow behavior of the private commitAll method, in case there are nested transactions.
+                    if (false === $connection->isAutoCommit() && 1 === $connection->getTransactionNestingLevel()) {
```

Comment:
I'm not too familiar with Doctrine, but if it's possible to run `$connection->commit();` before the 
`if (false === $connection->isAutoCommit() && 1 === $connection->getTransactionNestingLevel()) {`
It'd save a line of code, but I'm not sure if that'd make the `if` check invalid.

Reply:
This is a weird one, I'm trying to pretend to effectively run commitAll which is a private method that I can't call, so this while loop logic is effectively whats nested in there, just using public methods.

Honestly, though, I'm wondering if this is enough... What about flushing/persisting entity manager changes?